### PR TITLE
split #1351, libtestutils

### PR DESCRIPTION
**Pull request by debris**
Includes ClientBase . Please review #1407 first.

### CODE REVIEW

**Comment by chriseth** [comment]
Some documentation for the files themselves and the classes would be good.

**Comment by debris** [comment]
I added @brief documentation to `libtestutils` classes.

File: libtestutils/BlockChainLoader.cpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+using namespace dev::test;
+using namespace dev::eth;
+
+namespace dev
+{
+namespace test
+{
+dev::eth::BlockInfo toBlockInfo(Json::Value const& _json);
+bytes toGenesisBlock(Json::Value const& _json);
+
```
Comment (gavofyork) [inline libtestutils/BlockChainLoader.cpp]: either space before and after or neither

File: libtestutils/BlockChainLoader.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+/**
+ * @brief Should be used to load test blockchain from json file
+ * Loads the blockchain from json, creates temporary directory to store it, removes the directory on dealloc
+ */
+class BlockChainLoader
+{
+public:
+	BlockChainLoader(Json::Value const& _json);
+	eth::BlockChain& bc() { return *m_bc; }
+	eth::State state() { return m_state; }
```
Comment (gavofyork) [inline libtestutils/BlockChainLoader.h]: `const` method, and return `State const&`.

File: libtestutils/BlockChainLoader.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @brief Should be used to load test blockchain from json file
+ * Loads the blockchain from json, creates temporary directory to store it, removes the directory on dealloc
+ */
+class BlockChainLoader
+{
+public:
+	BlockChainLoader(Json::Value const& _json);
+	eth::BlockChain& bc() { return *m_bc; }
```
Comment (gavofyork) [inline libtestutils/BlockChainLoader.h]: can't return `BlockChain const&`? why is mutability needed?

File: libtestutils/Common.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+{
+
+struct TestChannel: public LogChannel  { static const char* name() { return "TEST"; } };
+#define ctest dev::LogOutputStream<dev::test::TestChannel, true>()
+
+std::string getTestPath();
+int randomNumber();
+Json::Value loadJsonFromFile(std::string const& _path);
+std::string toTestFilePath(std::string const& _filename);
+std::string getRandomPath();
```
Comment (gavofyork) [inline libtestutils/Common.h]: probably belongs in `libdevcrypto/FileSystem.h`

File: libtestutils/ShortLivingDirectory.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+{
+namespace test
+{
+
+/**
+ * @brief temporary directory implementation
+ * It creates temporary directory in the given path. On dealloc it removes the directory
+ * @throws if the given path already exists, throws an exception
+ */
+class ShortLivingDirectory
```
Comment (gavofyork) [inline libtestutils/ShortLivingDirectory.h]: `TransientDirectory`?

File: libtestutils/ShortLivingDirectory.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+{
+public:
+	ShortLivingDirectory(std::string const& _path = getRandomPath());
+	~ShortLivingDirectory();
+
+	std::string const& path() { return m_path; }
```
Comment (gavofyork) [inline libtestutils/ShortLivingDirectory.h]: don't forget your `const`s!

File: libtestutils/StateLoader.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+class StateLoader
+{
+public:
+	StateLoader(Json::Value const& _json);
+	eth::State const& state() const { return m_state; }
+
+private:
+	eth::State m_state;
+};
+}
```
Comment (gavofyork) [inline libtestutils/StateLoader.h]: blank line here

**Pull request merged**